File: 0637. Average of Levels in Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution(object):
    def averageOfLevels(self, root):
        if not root:
            return []

        queue = collections.deque([root])
        averages = []

        while queue:
            size = len(queue)
            level_sum = 0

            for i in range(size):
                node = queue.popleft()

                level_sum += node.val

                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

            averages.append (float(level_sum) / size)

        return averages

File: Java/7. Binary Tree/Traversal (Tree Level Order)/0637. Average of Levels in Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public List<Double> averageOfLevels(TreeNode root) {
        List<Double> resultList = new ArrayList<>(); // 用來裝答案
        Deque<TreeNode> que = new LinkedList<>(); // 輔助用的 Deque

        if (root == null) return resultList;
        que.offerLast(root); // 在進入 while loop 前先把當前 root 家進 queue

        while (!que.isEmpty()) {
            int levelSize = que.size(); // 這層有幾個 elements
            double levelSum = 0.0; // 加總先 initialized

            for (int i = 0; i < levelSize; i++) {
                TreeNode poll = que.pollFirst(); // 先噴出第一個
                levelSum += poll.val; // 加

                if (poll.left != null) que.addLast(poll.left);
                if (poll.right != null) que.addLast(poll.right);
            }
            resultList.add(levelSum / levelSize);
        }
        return resultList;
    }
}